extensions/amp-social-share/0.2/social-share.js

```diff
@@ -0,0 +1,188 @@
+      bindings[bindingName] = urlParams[name];
+    });
+  }
+
+  const {0: href, 1: setHref} = useState();
+  const {0: target, 1: setTarget} = useState();
+
+  urlReplacements.expandUrlAsync(hrefWithVars, bindings).then(result => {
+    setHref(result);
+    // mailto:, sms: protocols breaks when opened in _blank on iOS Safari
+    const {protocol} = Services.urlForDoc(props['host']).parse(href);
+    const isMailTo = protocol === 'mailto:';
+    const isSms = protocol === 'sms:';
+    setTarget(
+      platform.isIos() && (isMailTo || isSms)
+        ? '_top'
+        : props['dataTarget'] || '_blank'
+    );
+    if (isSms) {
+      // http://stackoverflow.com/a/19126326
+      // This code path seems to be stable for both iOS and Android.
+      setHref(href.replace('?', '?&'));
+    }
+  });
+
+  /**
+   * Handle key presses on the element.
+   * @param {!Event} event
+   * @private
+   */
+  function handleKeyPress(event) {
+    const {key} = event;
+    if (key == Keys.SPACE || key == Keys.ENTER) {
+      event.preventDefault();
+      handleActivation();
+    }
+  }
+
+  /**
+   * Handle clicks on the element.
+   * @private
+   */
+  function handleClick() {
+    handleActivation();
+  }
+
+  /** @private */
+  function handleActivation() {
+    userAssert(href && target, 'Clicked before href is set.');
+    dev().assertString(href);
+    dev().assertString(target);
+    if (shareEndpoint === 'navigator-share:') {
+      devAssert(navigator.share !== undefined, 'navigator.share disappeared.');
+      // navigator.share() fails 'gulp check-types' validation on Travis
+      navigator['share'](parseQueryString(href.substr(href.indexOf('?'))));
+    } else {
+      const windowFeatures = 'resizable,scrollbars,width=640,height=480';
+      openWindowDialog(window, href, target, windowFeatures);
+    }
+  }
+
+  const attrs = {
+    onClick: handleClick,
+    onKeydown: handleKeyPress,
+    className: `amp-social-share-${props['type']}`,
+    style: socialShareStyle + props['style'],
+    role: 'button',
+    tabindex: props['tabindex'] || '0',
+  };
+
+  const render = props['render'];
+  useResourcesNotify();
+  return render(null, props['children'], createElement('div', attrs));
+}
+
+/**
+ * @param {string=} type
+ * @param {!../../../src/service/viewer-interface.ViewerInterface} viewer
+ * @param {!../../../src/service/platform-impl.Platform} platform
+ * @return {dict=}
+ */
+function getTypeConfigOrUndefined(type, viewer, platform) {
+  userAssert(type, 'The type property is required. %s');
+  userAssert(
+    !/\s/.test(type),
+    'Space characters are not allowed in the type property. %s'
+  );
+  if (type === 'system') {
+    // // navigator.share unavailable
+    if (!systemShareSupported(viewer, platform)) {
+      return;
+    }
+  } else {
+    // system share wants to be unique
+    const systemOnly =
+      systemShareSupported(viewer, platform) &&
+      !!window.document.querySelectorAll(
```

Comment:
I don't think we should encourage DOM access here. Why not pass this as a prop? Or do the entire check in `init`?

Reply:
Good point, moved this logic to `init`